_drafts/hw_actors.md

```diff
@@ -0,0 +1,150 @@
+        }
+
+
+Queues may contain both messages and actors. If someone posts a message in an empty queue, it will save that message until some actor will try to poll. On the other hand, if an actor tries to get a message from an empty queue it will be saved as a subscriber and later when a message arrives the actor will be activated. Therefore, a queue at any moment of time may contain either messages or actors, but never both. Note that queue is just a head of linked list, it has no internal buffers, so queue overflow cannot happen.
+
+For simplicity, messages are fixed-sized. Message allocator is just a pool of fixed-sized memory chunks. For latency-critical systems, it may be even implemented as lock-free. When the message pointer arrives to an actor, it is assumed that message type is known to its recipient since it knows what queue it polls.
+
+
+## How it works
+
+
+As in the case of a thread scheduler, there is some global context containing runqueues, or ready-to-run lists. Each runqueue is mapped to some interrupt vector (unused by devices in that given application). Priorities of such vectors are adjusted to reflect runqueue priority.
+
+When some event (hardware interrupt) occurs, it allocates and posts a message to some queue, this causes activation of the subscribed actor, moving the message into its incoming mailbox, moving actor to the list of ready ones and also triggering interrupt vector corresponding to actor's priority. After leaving the interrupt handler CPU hardware automatically transfers control to the activated vector, its handler then calls framework's function 'schedule', which eventually calls the corresponding actor.
+During this process, the system remains fully preemptable and asynchronous: other interrupt handlers may post messages to other queues, if corresponding actors are delayed then messages just will be accumulated in queues. When some actor or interrupt activates another high-priority actor, then preemption occurs immediately, so this system has good response times and less jitter than loop-based solutions.
+
+Normally, the actor model does not require explicit synchronization like semaphores and mutexes because there are no mutable shared state. However, there may be shared resources, such as devices. In such a case mutual exclusion may be represented as multiple actors posting their requests to a single queue and one actor associated with that queue who will do all the work sequentially. Nevertheless, in practice, it is often desirable to block preemption. This may be done using hardware features for temporarily raising a running handler’s priority to some ceiling. Ability to block certain interrupt priorities without disabling interrupts completely also helps with response times since high-priority actors may never be blocked.
+
+Note that the interrupt vector is assigned to priority level, not to an actor, so the total number of actors is unlimited:
+
+![](/img/hw_actors/vectors.png)
```

Comment:
I'm not sure I follow the design entirely. So the run queues have multiple actors in them, and those actors themselves have message queues to receive messages?

Reply:
Hi, thank you for the feedback! I will try to fix all the issues you mentioned. Since fixing requires that parts of the article have to be rewritten it may take some time. So, I will update the PR once fixes will be ready. I have already used analog of grammarly as stated in submission manual (I experience issues with the grammarly itself) but may be this is still not enough.

Reply:
Actors have no internal storage (except the pointer to the message they’re handling), queues are separate objects that are not shown. Initially each actor is subscribed to some queue, since they are waiting they are not in the runqueues. Once someone pushes a message to the queue, the subscribed actor is moved to runqueue. The picture depicts a situation where multiple actors with different priorities are ready to run and therefore multiple vectors are pending. I hope I fixed it in the updated version.